zebra/dplane_fpm_nl.c

```diff
@@ -0,0 +1,665 @@
+		if (rv <= 0) {
+			zlog_debug("%s: netlink_route_multipath failed",
+				   __func__);
+			return 0;
+		}
+
+		nl_buf_len += (size_t)rv;
+		if (STREAM_WRITEABLE(fnc->obuf) < nl_buf_len) {
+			zlog_debug("%s: not enough output buffer (%ld vs %lu)",
+				   __func__, STREAM_WRITEABLE(fnc->obuf),
+				   nl_buf_len);
+			return -1;
+		}
+		break;
+
+	case DPLANE_OP_MAC_INSTALL:
+	case DPLANE_OP_MAC_DELETE:
+		rv = netlink_macfdb_update_ctx(ctx, nl_buf, sizeof(nl_buf));
+		if (rv <= 0) {
+			zlog_debug("%s: netlink_macfdb_update_ctx failed",
+				   __func__);
+			return 0;
+		}
+
+		nl_buf_len = (size_t)rv;
+		if (STREAM_WRITEABLE(fnc->obuf) < nl_buf_len) {
+			zlog_debug("%s: not enough output buffer (%ld vs %lu)",
+				   __func__, STREAM_WRITEABLE(fnc->obuf),
+				   nl_buf_len);
+			return -1;
+		}
+		break;
+
+	case DPLANE_OP_NH_INSTALL:
+	case DPLANE_OP_NH_UPDATE:
+	case DPLANE_OP_NH_DELETE:
+	case DPLANE_OP_LSP_INSTALL:
+	case DPLANE_OP_LSP_UPDATE:
+	case DPLANE_OP_LSP_DELETE:
+	case DPLANE_OP_PW_INSTALL:
+	case DPLANE_OP_PW_UNINSTALL:
+	case DPLANE_OP_ADDR_INSTALL:
+	case DPLANE_OP_ADDR_UNINSTALL:
+	case DPLANE_OP_NEIGH_INSTALL:
+	case DPLANE_OP_NEIGH_UPDATE:
+	case DPLANE_OP_NEIGH_DELETE:
+	case DPLANE_OP_VTEP_ADD:
+	case DPLANE_OP_VTEP_DELETE:
+	case DPLANE_OP_SYS_ROUTE_ADD:
+	case DPLANE_OP_SYS_ROUTE_DELETE:
+	case DPLANE_OP_ROUTE_NOTIFY:
+	case DPLANE_OP_LSP_NOTIFY:
+	case DPLANE_OP_NONE:
+		break;
+
+	default:
+		zlog_debug("%s: unhandled data plane message (%d) %s",
+			   __func__, dplane_ctx_get_op(ctx),
+			   dplane_op2str(dplane_ctx_get_op(ctx)));
+		break;
+	}
+
+	/* Skip empty enqueues. */
+	if (nl_buf_len == 0)
+		return 0;
+
+	/*
+	 * FPM header:
+	 * {
+	 *   version: 1 byte (always 1),
+	 *   type: 1 byte (1 for netlink, 2 protobuf),
+	 *   len: 2 bytes (network order),
+	 * }
+	 */
+	stream_putc(fnc->obuf, 1);
+	stream_putc(fnc->obuf, 1);
+	assert(nl_buf_len < UINT16_MAX);
+	stream_putw(fnc->obuf, nl_buf_len + 4);
+
+	/* Write current data. */
+	stream_write(fnc->obuf, nl_buf, (size_t)nl_buf_len);
+
+	/* Tell the thread to start writing. */
+	thread_add_write(fnc->fthread->master, fpm_write, fnc, fnc->socket,
+			 &fnc->t_write);
+
+	return 0;
+}
+
+/**
+ * Send all RIB installed routes to the connected data plane.
+ */
+static int fpm_rib_send(struct thread *t)
+{
+	struct fpm_nl_ctx *fnc = THREAD_ARG(t);
+	rib_dest_t *dest;
+	struct route_node *rn;
+	struct route_table *rt;
+	struct zebra_dplane_ctx *ctx;
+	rib_tables_iter_t rt_iter;
+
+	rt_iter.state = RIB_TABLES_ITER_S_INIT;
+	while ((rt = rib_tables_iter_next(&rt_iter))) {
+		for (rn = route_top(rt); rn; rn = srcdest_route_next(rn)) {
+			dest = rib_dest_from_rnode(rn);
+			/* Skip bad route entries. */
+			if (dest == NULL || dest->selected_fib == NULL) {
+				continue;
+			}
+
+			/* Check for already sent routes. */
+			if (CHECK_FLAG(dest->flags, RIB_DEST_UPDATE_FPM)) {
+				continue;
+			}
+
+			/* Enqueue route install. */
+			ctx = dplane_ctx_alloc();
```

Comment:
so let's come up with a way for you to re-use a context object in this loop, so we don't have to do so many alloc/free cycles. the code's all there to init / re-init the objectg - it's just a matter of making it available in the right way so it's easy for you to use?

Reply:
If you can manage it in a easy way it would be great. I did not want to touch this part because it seem to do more allocations in the opaque `struct zebra_dplane_ctx` and I didn't want to add more complexity to this PR.